### PR TITLE
Adds npm Lockfile v3 support (part A)

### PR DESCRIPTION
**Pull request by meghfossa**
# Overview

This PR adds initial lockfile v3 support. 
- We only apply this new tactic when we are confident that provided lockfile is v3 (we can parse, and lockFileVersion is explicitly 3)

Out of Scope:
- Nested workspaces (hope to address with part b of pr chain)
- Other resolutions work as before

## Acceptance criteria

- Npm lockfile v3 is supported

## Testing plan

1) Create an example: `package.json`

```
{
    "name": "lockv3",
    "version": "1.0.0",
    "description": "",
    "main": "index.js",
    "scripts": {
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "devDependencies": {
        "mocha": "*"
    },
    "peerDependencies": {
        "chalk": "*"
    },
    "dependencies": {
        "3": "^2.1.0",
        "aws-sdk": "^2.0.0",
        "react": "^18.1.0"
    },
    "author": "",
    "license": "ISC"
}

```

2) Execute: `npm i --lockfile-version=3`
3) Run `fossa-dev analyze -o` and confirm we analyzed it successfully.

## Manual Testing plan
- [x] compare v2 and v3 analysis for a complicated project
- [x] test against symbolic linked package

## To Dos
- [x] Update Changelog.md 

## Risks

Since lockFile v3 is not used by npm's default configs. This change has very low risk profile.

## References

https://fossa.atlassian.net/browse/ANE-109

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).

### CODE REVIEW

**Comment by meghfossa** [comment]
@csasarak tagging you here since you have lots of context in npm world.

**Comment by scruffystuffs** [comment]
Please block merge on my review as well.

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (meghfossa) [inline src/Strategy/Node/Npm/PackageLockV3.hs:437]: Idk how OK with these - from my experience, npm is complexly subtle that we may have a few defects after initial support work. I would like to keep something like this for easy debugging, but OK with killing it as well - will defer to review.

File: src/Strategy/Node.hs
Comment (meghfossa) [inline src/Strategy/Node.hs:225]: - [ ] TODO: Change it to `NpmLockV3` instead, as we don't run npmLock analysis for v2 lockfile

**Review by csasarak — CHANGES_REQUESTED** [review]
LGTM! Most of my comments were on pretty minor readability things, I don't expect you to do any of those if you don't want to. The one change I'm requesting here is the explicit setting of `EnvProduction` as an environment when building the dependencies.

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (csasarak) [inline src/Strategy/Node/Npm/PackageLockV3.hs:420]: Could the message here be "Analyzing Npm Lockfile (V3)"? The current output will say that we're parsing a v3 file, but it's not totally clear to me that it's using that data to build the graph in the output I see:
```
[DEBUG] [TASK 35] Project Analysis > Analyzing Npm Lockfile
[DEBUG] [TASK 35] Project Analysis > Analyzing Npm Lockfile > Parsing v3 compatible package-lock.json
[DEBUG] [TASK 35] Project Analysis > Analyzing Npm Lockfile > Parsing v3 compatible package-lock.json > Parsing JSON file 'foo/package-lock.json'
[ INFO] [ 0 Waiting / 3 Running / 33 Completed ]
[DEBUG] [TASK 35] Project Analysis > Analyzing Npm Lockfile > Building dependency graph

```
The only clue I see is that it's in close proximity to the parsing log messages. Or should the `Task 35` text be enough to group them together when reading?

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (meghfossa) [reply src/Strategy/Node/Npm/PackageLockV3.hs:69]: We shouldn't be hitting this point with an incompatible lockfile, as `Node.analyzeNpmLock` identifies the lockfile version and selects the appropriate tactic. When detecting the version, we are swallowing an error so - the user should not see any error noise.

Comment (csasarak) [reply src/Strategy/Node/Npm/PackageLockV3.hs:69]: Is this check primarily just for better error messages in the event that the higher level check goes away?

Comment (meghfossa) [reply src/Strategy/Node/Npm/PackageLockV3.hs:69]: Yep + as a guard, so we don't use `PkgLockV3` for non-v3 compatible fmt.

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:250]: [required]: Use `map` instead of `<$>` or `fmap` for lists:
```suggestion
              map toTopLevelPackage $
                concatMap
```

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:288]: [required]: Use `map` instead of `<$>` or `fmap` for lists:
```suggestion
              map (vendoredPathElseTopLevelPath workspaceRootPath) $
                concatMap
```

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:332]: [required]: Use `map` instead of `<$>` or `fmap` for lists:
```suggestion
              map (vendoredPathElseTopLevelPath vendorPathPrefix) $
                concatMap
```

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:345]: [nit]: no need for the lambda
```suggestion
            for_ resolvedDeepDeps $ edge parentDep
```

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:350]: [docs]: Can we add more comments making it clear how we handle a path like `workspace/node_modules/a/node_modules/b`?

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:374]: [nit]: add Line break
```suggestion
        _ -> Nothing -- don't report workspace or root project as dependency, or pkg not found

    allDependencyPackages :: Map PackagePath PackageLockV3Package
```

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [reply src/Strategy/Node/Npm/PackageLockV3.hs:425]: I'm fine with these, but let's add a `debugRepl` somewhere in the name, and annotate it with `{-# WARNING "some message" #-}` so that we can't accidentally use it.

We should make these easy to `grep` for, which is why I suggested `debugRepl`.  That way we can easily find them across the project for debugging.

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:417]: [nit]: We should generally avoid nested `where` clauses.  The `rootWorkspace` above is fine, but these don't depend on the `prefix` binding, so we can push them out to the top `where` clause to avoid cluttering up the nested one.
```suggestion
    getRootWorkspace :: Text -> Maybe Text
    getRootWorkspace pkgPath = case Text.breakOn "/node_modules" pkgPath of
      ("", "") -> Nothing
      (firstModule, _)
        | firstModule == pkgPath -> Nothing
        | (Map.size (Map.filterWithKey (\k _ -> isWorkSpace firstModule k) (packages pkgLockV3)) > 0) -> Just firstModule
        | otherwise -> Nothing

    isWorkSpace :: Text -> PackagePath -> Bool
    isWorkSpace name pkgPath = pkgPath == PackageLockV3WorkSpace name
```

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:422]: [org]: Can we move this function above `buildGraph`?

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:302]: [spelling]:
```suggestion
      -- "node_modules/someDirectDep": {
```

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:317]: [docs]: We should probably make it more clear in multiple places how we're handling paths like `workspace/node_modules/a/node_modules/b`.  I know that it's duplicating information, but it's a pretty critical detail that's not obvious from the code (mostly because this module is large enough that no minor detail is obvious until you read the whole thing).

This is one place where I'd like to see a direct call-out for sub-vendored packages within a workspace.

File: src/Strategy/Node/Npm/PackageLockV3.hs
Comment (scruffystuffs) [inline src/Strategy/Node/Npm/PackageLockV3.hs:409]: [nit]: Let's add a comment with example input/output here.

File: test/GraphUtil.hs
Comment (scruffystuffs) [inline test/GraphUtil.hs:64]: [nit]: Please also add `expectEdge'`.

File: test/Node/PackageLockV3Spec.hs
Comment (scruffystuffs) [inline test/Node/PackageLockV3Spec.hs:3]: [required]: Can we add tests for sub-vendored paths within workspaces?  e.g. `workspace/node_modules/a/node_modules/b`

**Review by scruffystuffs — CHANGES_REQUESTED** [review]
Lots of comments, I've marked the ones blocking approval as `[required]`.  Nothing wrong structurally, and the new module looks really clean and has great doc comments

**Review by scruffystuffs — APPROVED** [review]

**Comment by meghfossa** [comment]
@scruffystuffs resolved feedback with: https://github.com/fossas/fossa-cli/pull/932/commits/6f19f3613f23a6f86908a5fec1f343a4e5611358

**Review by csasarak — APPROVED** [review]

**Comment by thernstig** [comment]
@meghfossa I noticed this line:

> Nested workspaces (hope to address with part b of pr chain)

With workspaces in npm, one should only ever have a top-level package-lock.json file, and no package-lock.json files in the nested workspaces.

So what does that actual comment mean?

**Comment by meghfossa** [comment]
> With workspaces in npm, one should only ever have a top-level package-lock.json file, and no package-lock.json files in the nested workspaces.
> So what does that actual comment mean?

The point of interest (or scope for part B) is to address what happens when npm has resolved version conflicts for the same package across root (`aws-sdk: ^1.0.0`), workspace (`aws-sdk: ^2.0.0`), and nested workspace (`aws-sdk: ^3.00.0`). Although current implementation (per this PR) may work for such scenarios - I would like to ensure the correctness in secondary PR (via unit tests, and/or code change). 

I've noticed that npm-cli has had some defects reported in the resulting `package-lock.json` v8.x (e.g. https://github.com/npm/cli/issues/4774) when working against nested workspaces, which I would like to review before closing initial support for npm lockfile v3.

**Comment by thernstig** [comment]
@meghfossa please see https://github.com/fossas/fossa-cli/issues/940

**Pull request merged**
